File: day3/cpp/main.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

static int solve(const vector<string> lines, const int xinc, const int yinc)
{
    auto xp = 0;
    auto count = 0;
    const auto line_width = lines[0].size();
    auto y = yinc;
    while (y < lines.size())
    {
        xp += xinc;
        const auto actual_xp = xp % line_width;
        auto prev_line = lines[y];

        if (prev_line[actual_xp] == '#') count++;
        y += yinc;
    }

    return count;
}

int main()
{
    const string filename = "c:\\temp\\aoc3f.txt";
    ifstream inf(filename);
    vector<string> lines;
    string line;
    for (unsigned int i = 1; getline(inf, line); ++i)
        lines.emplace_back(line);
	
    cout << "SolvePart 1: " << solve(lines, 3, 1) << " trees\n";

    vector<int> nums;
    nums.emplace_back(solve(lines, 1, 1));
    nums.emplace_back(solve(lines, 3, 1));
    nums.emplace_back(solve(lines, 5, 1));
    nums.emplace_back(solve(lines, 7, 1));
    nums.emplace_back(solve(lines, 1, 2));
    auto sum = 1;
    for (auto num : nums) sum *= num;
    cout << "SolvePart 2: " << sum << "\n";
}

File: day3/swift/main.swift
// AoC day 3

import Foundation

extension String {
    subscript(offset: Int) -> Character {
        self[index(self.startIndex, offsetBy: offset)]
    }
}

func readFile(filename: String) -> [String] {
    do {
        let text = try String(contentsOfFile: filename, encoding: .ascii)
        let lines : [String] = text.components(separatedBy: "\n")
        return lines
    } catch let error as NSError{
        return[]
    }
}

func solve(lines: [String], xinc: Int, yinc: Int) -> Int {
    var xp = 0
    var count = 0
    let lineWidth = lines[0].trimmingCharacters(in: .whitespacesAndNewlines).count
    var y = yinc
    let lineCount = lines.count
    while(y < lineCount) {
        xp += xinc
        let actualXp = xp % lineWidth
        let prevLine = lines[y]
        if prevLine[actualXp] == "#" {
            count += 1
        }
        y += yinc
    }
    return count
}

let map = readFile(filename: "/Users/mikaelstalvik/Downloads/aoc3f.txt")

let res = solve(lines: map, xinc: 3, yinc: 1)

print("Solve part 1: Trees \(res)")

var nums: [Int] = []
nums.append(solve(lines: map, xinc: 1, yinc: 1))
nums.append(solve(lines: map, xinc: 3, yinc: 1))
nums.append(solve(lines: map, xinc: 5, yinc: 1))
nums.append(solve(lines: map, xinc: 7, yinc: 1))
nums.append(solve(lines: map, xinc: 1, yinc: 2))
var sum = 1
for num in nums {
    sum *= num
}
print(sum)

